### PR TITLE
Support different orientations for hexagonal lattices 

### PR DESCRIPTION
**Pull request by dryuri92**
I think will be helpful for OpenMC users to give opportunity to create diverse hexagonal lattice which oriented both along OY (as works in current program version ) and OX axis. My proposal is to add new field to hexlattice description called 'orient' not present by default OY long axis orientation and equal "ox" at an another one option. Description at geometry.xml will be like:
``
...
 <hex_lattice id="110" n_axial="1" n_rings="2" orient="OX" name="BN-1200 lattice">
    <pitch>... 
``
...
universes
    1  2 
   3  4  5 
    6  7 /universes

By default field 'orient' is absent. Coordinate system be like in OY orientation skewed but (alpha,y) fig 1 with alpha equal 60 degrees instead of (x,alpha). It allows to use the same indexation and avoid check range problem.  
ox_19hex
Fig 1. Skewed (alpha,y) coodinate system
On Python side new "OX" lattice be like:
```
lattices = {}
lattices['Reactor core'] = openmc.HexLattice(lattice_id=1, name='Reactor core lattice')
lattices['Reactor core'].orientation = "OX"

```
when the "OY" be write the same way as early. Numeration in python are some different for "OX" from "OY" case and opposite clockwise but ring and axial layer the same.
 ```
          (1, 2)      (1, 1)      

    (1, 3)      (2, 0)      (1, 0)      

           (1, 4)      (1, 5)  
```
For what ? It exists a lot of problem and benchmarks (like Russian's VVER, research and contemporary fast reactors https://www.oecd-nea.org/science/docs/2015/)  with hexagonal lattice when pins in assembly oriented along OY axis at the same time assembly in the  reactor core lattice along OX.  My problem plotting by OpenMC tools looks like
image
Fig 2. 7 assemblies in core with 127 pins
I ve been tested on my issues and get the same result in value of keff and tallies with different orientation for standalone hexagonal with equal reflective boundary conditions. If my help to modify code or fit to your standart or add some tests is necessary  ,I try to correct it.

### CODE REVIEW

**Comment by paulromano** [comment]
Thanks for the contribution @dryuri92! I should have some time to take a closer look within the next day or two.

**Comment by paulromano** [comment]
So one thing we should discuss before getting too much into the details -- the notation that @dryuri92 is using ("OY" meaning two sides are y-planes and "OX" meaning two sides are x-planes) is consistent with the notation that Serpent uses (hexxc and hexyc) but is inconsistent with the currently existing [`get_hexagonal_prism`](https://openmc.readthedocs.io/en/latest/pythonapi/generated/openmc.model.get_hexagonal_prism.html#openmc.model.get_hexagonal_prism) function in our Python API. We should either

1. Make the orientation for our hexagonal lattices consistent with `get_hexagonal_prism`, or
2. Keep it consistent with Serpent's convention and modify `get_hexagonal_prism`.

If we go with option 2, I think the best thing to do is actually to keep `get_hexagonal_prism` as is so we don't break existing scripts but introduce a new function `hexagonal_prism` that has the new convention. I've been wanting to make such a name change anyway.

Any thoughts @openmc-dev/committers?

**Comment by paulromano** [comment]
@dryuri92 There are a few small conflicts with the `develop` branch that you need to resolve. Assuming you have a remote called `upstream` that corresponds to the main OpenMC repo, you could resolve the conflicts with something like:
```
git fetch upstream
git merge upstream/develop
... resolve any conflicts by editing the files ...
git add <conflicted files>
git commit
```
Let me know if you have any questions about how to handle the conflicts.

Also, another thing I noticed is that the lines of code that you added use tab characters. The codebase currently does not use tab characters and instead uses spaces for indentation. Please convert the tabs to spaces to stay consistent.

If you haven't yet, please take a look at our [style guide](https://openmc.readthedocs.io/en/latest/devguide/styleguide.html) for the project. Once again, thanks for your contribution and welcome to the community! I'll provide some more detailed feedback on the code soon.

**Comment by dryuri92** [comment]
> @dryuri92 There are a few small conflicts with the `develop` branch that you need to resolve. Assuming you have a remote called `upstream` that corresponds to the main OpenMC repo, you could resolve the conflicts with something like:
> 
> ```
> git fetch upstream
> git merge upstream/develop
> ... resolve any conflicts by editing the files ...
> git add <conflicted files>
> git commit
> ```
> 
> Let me know if you have any questions about how to handle the conflicts.
> 
> Also, another thing I noticed is that the lines of code that you added use tab characters. The codebase currently does not use tab characters and instead uses spaces for indentation. Please convert the tabs to spaces to stay consistent.
> 
> If you haven't yet, please take a look at our [style guide](https://openmc.readthedocs.io/en/latest/devguide/styleguide.html) for the project. Once again, thanks for your contribution and welcome to the community! I'll provide some more detailed feedback on the code soon.
@paulromano I ve replaced tabs on spaces and code tests on travis ci been passed successfully. About conflict in source file lattice.cpp. I understand the explicit declaration of named coordinate indexes in function get_indices() here:
```
int ix = std::floor(r_o.x / (0.5*std::sqrt(3.0) * pitch_[0]));
int ia = std::floor(alpha / pitch_[0]);
```
 is made to improve code readability, but is not good for my approach with (ix,ialpha) named coordinates for OY orientation skewed coordinate, and (ialpha,iy) named coordinates. We have in this function the identical algorythm finding cell coordinate  should work both with ix,ialpha and ialpha,iy declaration to avoid a mess. Thus  text of the  function will be expanded by several condition and declaration blocks. My proposal to use unnamed indexes i1,i2 which means (ix,ialpha) in first case and (ialpha,iy) in second one. Im not really insist on such solution and can add several strings if needed.

**Comment by smharper** [comment]
Hi @dryuri92.  Thanks for putting in the work on this!  I apologize for not reviewing sooner.

This feature is useful, but we should try reducing the amount of duplicated code between the two lattice orientations.  The distance function is a very big function to duplicate given that the calculations only differ by a 30-degree rotation.  Instead of having two functions, can we just have one function with a rotation block at the beginning like:
```C++
  if (hextype == 1) {
    const double SQRT3_2 {std::sqrt(3) * 0.5};
    r = Position(SQRT3_2*r.x - 0.5*r.y, 0.5*r.x + SQRT3_2*r.y, r.z)
    u = Direction(SQRT3_2*u.x - 0.5*u.y, 0.5*u.x + SQRT3_2*u.y, u.z)
  }
```

**Comment by dryuri92** [comment]
@smharper Sorry but I don't  understand you. Did you ask me for rewriting lattice direction function taking into account rotation angle of cell? - Yes I can do it, but now as I understand @paulromano making some changes to the same function and i don't want to duplicate work. Or maybe do you want to keep my approach with reducing some parts of code?

**Comment by smharper** [comment]
@dryuri92, it is possible we can merge both #1232 and this PR.  Are you happy to use the #1232 approach alone?  If so, then there is no need to continue making changes to this PR.  I just wanted to offer a path forward in case you felt strongly about the approach you used.

**Comment by dryuri92** [comment]
@smharper Yes, I think for user it will be helpful to select orientation of  hex lattice. If you've decided to keep this ability, I will try to cut off redundant part of code soon as you mentioned above.

**Comment by smharper** [comment]
That change removed 80 lines of code which is a good step, but there is more room for improvement.

Currently there are some very big differences between the basis vectors used in the OX and OY implementations.  This makes it a little harder to understand the algorithms, and it forces the use of the `translation_matrix`.  It would be much more clear if all of the OX vectors were simply -30 degree rotations of the OY vectors.

In OY, we have the (x, alpha) basis set which I will call (basis0, basis1) right now for clarity.  Then there are the beta, gamma, and y vectors which indicate the directions to the flat sides of each hexagonal cell.  I'll call them beta, gamma, and delta for now.  This gives us the set of vectors:
```
OY:
  basis0 = (1, 0)
  basis1 = (-1/sqrt(3), 1)   = +120 degrees from basis0
  beta   = (sqrt(3)/2, 1/2)  = +30 degrees from basis0
  gamma  = (sqrt(3)/2, -1/2) = -60 degrees from beta
  delta  = (0, 1)            = +60 degrees from beta
```

In the current state of this PR, the OX vectors are
```
Current OX:
  basis0 = (sqrt(3), 1)
  basis1 = (0, 1)            = -60 degrees from basis0
  beta   = (1/2, sqrt(3)/2)  = +90 degrees from basis0
  gamma  = (1/2, -sqrt(3)/2) = -120 degrees from beta
  delta  = (1, 0)            = -60 degrees from beta
```
Notice how the relative orientation of each vector is different than with OY.  I would like to propose instead:
```
Proposed OX:
  basis0 = (1/sqrt(3), -1)
  basis1 = (0, 1)            = +120 degrees from basis0
  beta   = (1, 0)            = +30 degrees from basis0
  gamma  = (1/2, -sqrt(3)/2) = -60 degrees from beta
  delta  = (1/2, sqrt(3)/2)  = +60 degrees from beta
```

That way, moving in the beta direction will result in a (1, 0, 0) lattice translation and moving in the delta direction will result in a (0, 1, 0) translation---just like with OY.

**Comment by dryuri92** [comment]
Sorry for delaying,I corrected code as you describe above.  If you have any consideration or proposal, I will modify code to fit it. My question about hdf5-file is actual.

**Comment by smharper** [comment]
The coding style (whitespace, placement of braces, etc.) deviates significantly from our style guide.  I've made a PR to your repository to show how the C++ should be styled.  I'll check the Python style later.

Your implementation of the HDF5 issue is perfect.  It's backwards compatible because HDF5 lattice entries that do not specify an orientation assume the old OY.  There is no need to update the statepoint version number or anything like that.

We should also add a regression test to make sure that we do not break the OX feature in the future.  I recommend copying the tests/regression_tests/lattice_hex test and switching it to an OX lattice.

**Comment by paulromano** [comment]
Thanks @dryuri92 for the revisions and @smharper for helping out. Just a heads up - I'd like to take a pass at reviewing this PR before we merge but won't be able to get to it until tomorrow most likely.

File: include/openmc/constants.h [hunk 1]
```diff
@@ -25,7 +25,7 @@ constexpr int VERSION_RELEASE {0};
 constexpr std::array<int, 3> VERSION {VERSION_MAJOR, VERSION_MINOR, VERSION_RELEASE};
 
 // HDF5 data format
-constexpr int HDF5_VERSION[] {2, 0};
+constexpr int HDF5_VERSION[] {2, 0};//DR change version HDF5 to work
```
Comment (smharper) [inline include/openmc/constants.h]: Remove this comment.  We don't need to update the version number.  (And this version number is for the nuclear data which is unaffected by the lattice changes.)

File: openmc/lattice.py [hunk 2]
```diff
@@ -1059,15 +1139,27 @@ def find_element(self, point):
...
+        if self._hextype:#//DR
+           alpha = y - x*sqrt(3.)
+           ix = 0
+           ia = floor(-alpha/(sqrt(3.0) * self.pitch[0]))
+           iy = floor(y/(sqrt(0.75) * self.pitch[0]))
+        else:
+           alpha = y - x/sqrt(3.)
+           ix = floor(x/(sqrt(0.75) * self.pitch[0]))
+           ia = floor(alpha/self.pitch[0])
+           iy = 0
```
Comment (smharper) [inline openmc/lattice.py]: It would be better to use `i1, i2` instead of `ix/ia, ia/iy` like you did in `HexLattice::get_indices` in lattice.cpp.  That will reduce the amount of code needed here, and it's always better for the Python code to match the C++ code as much as possible.

Indent 4 spaces rather than 3.

File: openmc/lattice.py [hunk 3]
```diff
@@ -1094,16 +1187,40 @@ def get_local_coordinates(self, point, idx):
...
 
         """
-        x = point[0] - (self.center[0] + sqrt(0.75)*self.pitch[0]*idx[0])
-        y = point[1] - (self.center[1] + (0.5*idx[0] + idx[1])*self.pitch[0])
+        if self._hextype:#//DR
+           x = point[0] - (self.center[0] + self.pitch[0]*idx[0] + 0.5*self.pitch[0]*idx[1])
+           y = point[1] - (self.center[1] + (sqrt(0.75)*self.pitch[0]*idx[1]))
+        else:
+           x = point[0] - (self.center[0] + sqrt(0.75)*self.pitch[0]*idx[0])
+           y = point[1] - (self.center[1] + (0.5*idx[0] + idx[1])*self.pitch[0])
```
Comment (smharper) [inline openmc/lattice.py]: Indent 4 spaces rather than 2.

File: openmc/lattice.py [hunk 4]
```diff
@@ -1094,16 +1187,40 @@ def get_local_coordinates(self, point, idx):
             system
 
         """
-        x = point[0] - (self.center[0] + sqrt(0.75)*self.pitch[0]*idx[0])
-        y = point[1] - (self.center[1] + (0.5*idx[0] + idx[1])*self.pitch[0])
+        if self._hextype:#//DR
+           x = point[0] - (self.center[0] + self.pitch[0]*idx[0] + 0.5*self.pitch[0]*idx[1])
```
Comment (smharper) [inline openmc/lattice.py]: Do something like
```python
           x = point[0] - (self.center[0] + self.pitch[0]*idx[0]
                           + 0.5*self.pitch[0]*idx[1])
```
to make the line shorter than 80 columns.

File: openmc/lattice.py [hunk 5]
```diff
@@ -1094,16 +1187,40 @@ def get_local_coordinates(self, point, idx):
...
+           x = point[0] - (self.center[0] + sqrt(0.75)*self.pitch[0]*idx[0])
+           y = point[1] - (self.center[1] + (0.5*idx[0] + idx[1])*self.pitch[0])
+
         if self._num_axial is None:
             z = point[2]
         else:
             z = point[2] - (self.center[2] + (idx[2] + 0.5 - 0.5*self.num_axial)*
                             self.pitch[1])
         return (x, y, z)
+    def get_universe_index(self, idx): #//DR
```
Comment (smharper) [inline openmc/lattice.py]: Put a blank line before this function.

File: openmc/lattice.py [hunk 6]
```diff
@@ -1094,16 +1187,40 @@ def get_local_coordinates(self, point, idx):
...
+        -------
+        2- or 3-tuple of int
+            Indices used when setting the :attr:`HexLattice.universes` property
+
+        """
+        if self._hextype:
+           return self.get_universe_index_ox(idx)
+        else:
+           return self.get_universe_index_oy(idx)
+    def get_universe_index_oy(self, idx):#//DR
```
Comment (smharper) [inline openmc/lattice.py]: Put a blank line before this function.

This function should never be called by the user (they'll use get_universe_index instead) so make it and the ox version private by prepending and underscore to the name like `def _get_universe_index_oy`

File: openmc/lattice.py [hunk 7]
```diff
@@ -136,11 +136,18 @@ def from_hdf5(group, universes):
...
         elif lattice_type == 'hexagonal':
+
             n_rings = group['n_rings'][()]
             n_axial = group['n_axial'][()]
             center = group['center'][()]
             pitch = group['pitch'][()]
             outer = group['outer'][()]
+            #DR Temp added for compatibility with previous version
+            if ('orientation' in group.keys()):
+               orientation = group['orientation'].value
```
[the remark this answers is not in the document]
Comment (dryuri92) [reply openmc/lattice.py]: Is it right  to use decode method this to transform `class <bytes>` to `class <string>`
```suggestion
               orientation = group['orientation'][()].decode()
```
 or is it possible to get a string and then to get  error lilke " not supporting method decode by string?

**Comment by dryuri92** [comment]
Thank you for help to fit code to required coding style! I've removed garbage and created new regression test with OX-orient. lattice.

File: openmc/lattice.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (paulromano) [reply openmc/lattice.py]: Yes, calling `.decode()` is the right way to convert the `bytes` object into a `str`.

File: openmc/lattice.py [hunk 8]
```diff
@@ -869,11 +934,13 @@ def __init__(self, lattice_id=None, name=''):
         self._num_rings = None
         self._num_axial = None
         self._center = None
+        self._hextype = 0 # by Default orientaion is OY
```
Comment (paulromano) [inline openmc/lattice.py]: I find the use of 0 and 1 to be a bit confusing. I would just use a string and call this attribute `_orientation`:
```Python
self._orientation = 'y'
```

This will require other changes anywhere `self._hextype` is used.

File: openmc/lattice.py [hunk 9]
```diff
@@ -901,7 +968,14 @@ def __repr__(self):
...
     def num_rings(self):
         return self._num_rings
-
+    
+    @property
+    def orientation(self):
+        if self._hextype:
+           return "OX"
+        else:
+           return "OY"
```
Comment (paulromano) [inline openmc/lattice.py]: With the changes suggested above, this would become:
```Python
return self._orientation
```

File: openmc/lattice.py [hunk 10]
```diff
@@ -955,6 +1029,12 @@ def center(self, center):
         cv.check_length('lattice center', center, 2, 3)
         self._center = center
 
+    @orientation.setter
+    def orientation(self, orientation):
+        cv.check_value('orientation', orientation.lower(), ('ox', 'oy'))
+        if orientation.lower() == "ox":
+           self._hextype = 1
```
Comment (paulromano) [inline openmc/lattice.py]:
```Python
self._orientation = orientation.lower()
```

File: openmc/lattice.py [hunk 11]
```diff
@@ -1045,7 +1125,7 @@ def find_element(self, point):
         -------
         3-tuple of int
             Indices of corresponding lattice element in :math:`(x,\alpha,z)`
-            bases
+            or :math:`(\alpha,y,z)`bases
```
Comment (paulromano) [inline openmc/lattice.py]: add space before bases

File: openmc/lattice.py [hunk 12]
```diff
@@ -1059,15 +1139,20 @@ def find_element(self, point):
         else:
             z = point[2] - self.center[2]
             iz = floor(z/self.pitch[1] + 0.5*self.num_axial)
-        alpha = y - x/sqrt(3.)
-        ix = floor(x/(sqrt(0.75) * self.pitch[0]))
-        ia = floor(alpha/self.pitch[0])
-
+        if self._hextype:
```
Comment (paulromano) [inline openmc/lattice.py]:
```Python
if self._orientation == 'x':
```

File: openmc/lattice.py [hunk 13]
```diff
@@ -1059,15 +1139,20 @@ def find_element(self, point):
...
+            i1 = floor(-alpha/(sqrt(3.0) * self.pitch[0]))
+            i2 = floor(y/(sqrt(0.75) * self.pitch[0]))
+        else:
+            alpha = y - x/sqrt(3.)
+            i1 = floor(x/(sqrt(0.75) * self.pitch[0]))
+            i2 = floor(alpha/self.pitch[0])
         # Check four lattice elements to see which one is closest based on local
         # coordinates
+        numbersaxy=[(i1, i2, iz), (i1 + 1, i2, iz), (i1, i2 + 1, iz),
+                 (i1 + 1, i2 + 1, iz)]
```
Comment (paulromano) [inline openmc/lattice.py]: For Python code, our [style guide](http://docs.openmc.org/en/latest/devguide/styleguide.html#python) says you need to follow [PEP8](https://www.python.org/dev/peps/pep-0008/). One rule there is that you should have a space around `=` in an assignment. It also says continuation lines should be lined up based on context above. I'd also change the variable name here to `indices`:
```Python
indices = [(i1, i2, iz), (i1 + 1, i2, iz), (i1, i2 + 1, iz),
           (i1 + 1, i2 + 1, iz)]
```

File: openmc/lattice.py [hunk 14]
```diff
@@ -1085,7 +1170,8 @@ def get_local_coordinates(self, point, idx):
         point : Iterable of float
             Cartesian coordinates of point
         idx : Iterable of int
-            Indices of lattice element in :math:`(x,\alpha,z)` bases
+            Indices of lattice element in :math:`(x,\alpha,z)` 
+            or :math:`(\alpha,y,z)` basesis
```
Comment (paulromano) [inline openmc/lattice.py]: typo: bases

File: openmc/lattice.py [hunk 15]
```diff
@@ -1094,18 +1180,48 @@ def get_local_coordinates(self, point, idx):
             system
 
         """
-        x = point[0] - (self.center[0] + sqrt(0.75)*self.pitch[0]*idx[0])
-        y = point[1] - (self.center[1] + (0.5*idx[0] + idx[1])*self.pitch[0])
+        if self._hextype:
+            x = point[0] - (self.center[0] + self.pitch[0]*idx[0] + 
+                            0.5*self.pitch[0]*idx[1])
+            y = point[1] - (self.center[1] + 
+                            (sqrt(0.75)*self.pitch[0]*idx[1]))
```
Comment (paulromano) [inline openmc/lattice.py]: extra set of parentheses here is not needed

File: openmc/lattice.py [hunk 16]
```diff
@@ -1094,18 +1180,48 @@ def get_local_coordinates(self, point, idx):
...
+                            + (0.5*idx[0] + idx[1])*self.pitch[0])
+
         if self._num_axial is None:
             z = point[2]
         else:
             z = point[2] - (self.center[2] + (idx[2] + 0.5 - 0.5*self.num_axial)*
                             self.pitch[1])
         return (x, y, z)
-
+
```
Comment (paulromano) [inline openmc/lattice.py]: Remove trailing whitespace (I'd recommend configuring your text editor to do this by default)

File: openmc/lattice.py [hunk 17]
```diff
@@ -1094,18 +1180,48 @@ def get_local_coordinates(self, point, idx):
...
+            system in 'OX' one 
+
+        Returns
+        -------
+        2- or 3-tuple of int
+            Indices used when setting the :attr:`HexLattice.universes` property
+
+        """
+        if self._hextype:
+           return self._get_universe_index_ox(idx)
```
Comment (paulromano) [inline openmc/lattice.py]: PEP8: four spaces per indentation level (need this on else block too)

File: openmc/lattice.py [hunk 18]
```diff
@@ -1143,13 +1259,55 @@ def get_universe_index(self, idx):
...
+        else:
+            if a <= 0:
+                i_within = 3*g - y
+            else:
+                i_within = 5*g - z
+
+        if self.num_axial is None:
+            return (i_ring, i_within)
+        else:
+            return (idx[2], i_ring, i_within)
```
Comment (paulromano) [inline openmc/lattice.py]: This function is almost identical to the y case. I would combine them and just conditionally set (x, a) and (a, y) depending on the orientation type. Perhaps you can call them a1 and a2 (as in axis 1 and axis 2). Something like:
```Python
if self._orientation == "x":
    a1, a2 = idx[0], idx[1]
else:
    a1, a2 = idx[1], idx[0]
```

File: openmc/lattice.py [hunk 19]
```diff
@@ -1193,6 +1351,9 @@ def create_xml_subelement(self, xml_element):
             self._outer.create_xml_subelement(xml_element)
 
         lattice_subelement.set("n_rings", str(self._num_rings))
+        # If orientation is "OX" export it to XML
+        if self._hextype:
+            lattice_subelement.set("orient", "OX")
```
Comment (paulromano) [inline openmc/lattice.py]: Let's use "orientation" to be consistent rather than "orient"

File: openmc/lattice.py [hunk 20]
```diff
@@ -1315,10 +1476,123 @@ def from_xml_element(cls, elem, get_universe):
...
+
+                # Translate the indices.
+                y += 1
+                theta += 1
+
+            # Climb left the top-left.
+            for i in range(r):
+                # Add the universe.
+                universe = universes[r_prime][theta]
+                rows[y].insert(0,id_form.format(universe._id))
```
Comment (paulromano) [inline openmc/lattice.py]:
```suggestion
                rows[y].insert(0, id_form.format(universe._id))
```

File: openmc/lattice.py [hunk 21]
```diff
@@ -1315,10 +1476,123 @@ def from_xml_element(cls, elem, get_universe):
...
+
+                # Translate the indices.
+                theta += 1
+
+            # Climb down the middle left.
+            for i in range(r):
+                # Add the universe.
+                universe = universes[r_prime][theta]
+                rows[y].insert(0,id_form.format(universe._id))
```
Comment (paulromano) [inline openmc/lattice.py]:
```suggestion
                rows[y].insert(0, id_form.format(universe._id))
```

File: openmc/lattice.py [hunk 22]
```diff
@@ -1315,10 +1476,123 @@ def from_xml_element(cls, elem, get_universe):
...
         """Return string representation for the given 2D group of universes.
 
+        The 'universes' argument should be a list of lists of universes where
+        each sub-list represents a single ring.  The first list should be the
+        outer ring.
+        """
+        if self._hextype:
+           return self._repr_axial_slice_ox(universes)
+        else:
+           return self._repr_axial_slice_oy(universes)
```
Comment (paulromano) [inline openmc/lattice.py]:
```suggestion
            return self._repr_axial_slice_oy(universes)
```

File: openmc/lattice.py [hunk 23]
```diff
@@ -1315,10 +1476,123 @@ def from_xml_element(cls, elem, get_universe):
...
-
     def _repr_axial_slice(self, universes):
         """Return string representation for the given 2D group of universes.
 
+        The 'universes' argument should be a list of lists of universes where
+        each sub-list represents a single ring.  The first list should be the
+        outer ring.
+        """
+        if self._hextype:
+           return self._repr_axial_slice_ox(universes)
```
Comment (paulromano) [inline openmc/lattice.py]: Need four spaces for indentation
```suggestion
            return self._repr_axial_slice_ox(universes)
```

File: openmc/lattice.py [hunk 24]
```diff
@@ -1315,10 +1476,123 @@ def from_xml_element(cls, elem, get_universe):
...
+        # Pad the beginning of the rows so they line up properly.
+        for y in range(self._num_rings - 1):
+            rows[y] = (self._num_rings - 1 - y)*pad + rows[y]
+            rows[-1 - y] = (self._num_rings - 1 - y)*pad + rows[-1 - y]
+
+        # Join the rows together and return the string.
+        universe_ids = '\n'.join(rows)
+        return universe_ids
+
+
```
Comment (paulromano) [inline openmc/lattice.py]: PEP8: One blank line between methods in a class

File: openmc/lattice.py [hunk 25]
```diff
@@ -1527,3 +1801,100 @@ def show_indices(num_rings):
...
+        This method can be used to show the proper indices to be used when
+        setting the :attr:`HexLattice.universes` property. For example, running
+        this method with num_rings=3 will return the similar diagram::
+
+             (0, 4)      (0, 3)      (0, 2)
+
+       (0, 5)      (1, 2)      (1, 1)      (0, 1)
+
+ (0, 6)      (1, 3)      (2, 0)      (1, 0)      (0, 0)
```
Comment (paulromano) [inline openmc/lattice.py]: This diagram needs to match the indentation of the rest of the docstring.

File: openmc/lattice.py [hunk 26]
```diff
@@ -1527,3 +1801,100 @@ def show_indices(num_rings):
 
         # Join the rows together and return the string.
         return '\n'.join(rows)
+    @staticmethod
```
Comment (paulromano) [inline openmc/lattice.py]: Add blank line above

File: openmc/lattice.py [hunk 27]
```diff
@@ -1527,3 +1801,100 @@ def show_indices(num_rings):
...
+    def show_indices_ox(num_rings):
```
Comment (paulromano) [inline openmc/lattice.py]: I would have a function:
```Python
def show_indices(num_rings, orientation="y"):
```
Rather than two different functions. It's ok to maintain the two logic as two internal functions (`_show_indices_y(num_rings)` and `_show_indices_x(num_rings)`)

File: src/initialize.cpp [hunk 28]
```diff
@@ -37,6 +37,10 @@ int openmc_init(int argc, char* argv[], const void* intracomm)
 {
   using namespace openmc;
 
+
+  char cCurrentPath[256];
+
+
```
Comment (paulromano) [inline src/initialize.cpp]: Remove these lines

File: src/lattice.cpp [hunk 29]
```diff
@@ -482,14 +497,83 @@ HexLattice::HexLattice(pugi::xml_node lat_node)
...
 
   universes_.resize((2*n_rings_-1) * (2*n_rings_-1) * n_axial_, C_NONE);
+
+  if (orientation_ == HexOrientation::oy) {
+    fill_lattice_oy(univ_words);
+  } else {
+    fill_lattice_ox(univ_words);
+  }
+
+
```
Comment (paulromano) [inline src/lattice.cpp]: Remove extra blank lines here

File: include/openmc/lattice.h [hunk 30]
```diff
@@ -265,8 +269,15 @@ class HexLattice : public Lattice
   void to_hdf5_inner(hid_t group_id) const;
 
 private:
+  //! Fill universes_ vector for OY orientation
+  void fill_lattice_oy(std::vector<std::string> univ_words);
```
Comment (paulromano) [inline include/openmc/lattice.h]: Passing a vector by value will result in it being copied. Instead, pass a constant reference:
```C++
void fill_lattice_y(const std::vector<std::string>& univ_words);
```

File: include/openmc/lattice.h [hunk 31]
```diff
@@ -265,8 +269,15 @@ class HexLattice : public Lattice
...
+
+  //! Fill universes_ vector for OX orientation
+  void fill_lattice_ox(std::vector<std::string> univ_words);
```
Comment (paulromano) [inline include/openmc/lattice.h]: Same thing here, use const reference

File: src/lattice.cpp [hunk 32]
```diff
@@ -797,13 +930,27 @@ Position
...
+  } else {
+    // x_l = x_g - (center + pitch_x*index_a + pitch_y*sin(30)*index_y)
+    r.x -= (center_.x + (i_xyz[0] - n_rings_ + 1) * pitch_[0]
+            + (i_xyz[1] - n_rings_ + 1) * pitch_[0] / 2.0);
+    // y_l = y_g - (center + pitch_y*cos(30)*index_y)
+    r.y -= center_.y
+           + std::sqrt(3.0)/2.0 * (i_xyz[1] - n_rings_ + 1) * pitch_[0];
+    if (is_3d_) {
+      r.z -= center_.z - (0.5 * n_axial_ - i_xyz[2] - 0.5) * pitch_[1];
+    }
```
Comment (paulromano) [inline src/lattice.cpp]: The `is_3d_` if block is the same for both orientations, so it can be moved outside of that if block.

**Comment by dryuri92** [comment]
@paulromano i'm sorry for  delayment. If some code improvement is needed i'm glad to fix or add some.

**Comment by smharper** [comment]
Looks good!  Thanks again, @dryuri92, for contributing
